cmd/registry/cmd/upload-csv_test.go

```diff
@@ -0,0 +1,232 @@
+// Copyright 2021 Google LLC. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cmd
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"testing"
+
+	"github.com/apigee/registry/connection"
+	"github.com/apigee/registry/rpc"
+	"github.com/apigee/registry/tests/seeding/fileseed"
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
+	"google.golang.org/api/iterator"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/testing/protocmp"
+)
+
+const (
+	gzipOpenAPIv3 = "application/x.openapi+gzip;version=3.0.0"
+	testProject   = "googleapis.com"
+)
+
+var (
+	specContents1 = []byte(`{"openapi": "3.0.0", "info": {"title": "First Spec", "version": "v1"}, "paths": {}}`)
+	specContents2 = []byte(`{"openapi": "3.0.0", "info": {"title": "Second Spec", "version": "v1"}, "paths": {}}`)
+	specContents3 = []byte(`{"openapi": "3.0.0", "info": {"title": "Third Spec", "version": "v1"}, "paths": {}}`)
+	specContents4 = []byte(`{"openapi": "3.0.0", "info": {"title": "Fourth Spec", "version": "v1"}, "paths": {}}`)
+)
+
+func TestUploadCSV(t *testing.T) {
+	root := t.TempDir()
+	tests := []struct {
+		desc  string
+		files []fileseed.File
+		args  []string
+		want  []*rpc.ApiSpec
+		opts  cmp.Option
+	}{
+		{
+			desc: "multiple spec upload",
+			files: []fileseed.File{
+				{
+					Path:     fmt.Sprintf("%s/cloudtasks/v2beta2/openapi.yaml", root),
+					Contents: specContents1,
+				},
+				{
+					Path:     fmt.Sprintf("%s/cloudtasks/v2/openapi.yaml", root),
+					Contents: specContents2,
+				},
+				{
+					Path:     fmt.Sprintf("%s/datastore/v1beta1/openapi.yaml", root),
+					Contents: specContents3,
+				},
+				{
+					Path:     fmt.Sprintf("%s/datastore/v1/openapi.yaml", root),
+					Contents: specContents4,
+				},
+				{
+					Path: fmt.Sprintf("%s/specs.csv", root),
+					Contents: []byte(strings.Join([]string{
+						"api_id,version_id,spec_id,filepath",
+						fmt.Sprintf("cloudtasks,v2beta2,openapi.yaml,%s/cloudtasks/v2beta2/openapi.yaml", root),
+						fmt.Sprintf("cloudtasks,v2,openapi.yaml,%s/cloudtasks/v2/openapi.yaml", root),
+						fmt.Sprintf("datastore,v1beta1,openapi.yaml,%s/datastore/v1beta1/openapi.yaml", root),
+						fmt.Sprintf("datastore,v1,openapi.yaml,%s/datastore/v1/openapi.yaml", root),
+					}, "\n")),
+				},
+			},
```

Comment:
After setting up these tests to write the file contents to a temporary directory as part of their setup I think I prefer the option of having fixed files in a test data directory.

If we're taking that approach I think it would be good to reconsider the location of cobra command tests. Since we're able to execute against the root command with provided arguments I think we can have an external test package that lives in a sibling directory like `cmd/registry/test`.

Reply:
I think reorganizing in a later PR is the way to go. That way it's one non-functional change instead of a mix of reorganizing and functional additions.